pkg/clustermesh/common/namespacewatcher/watcher.go

```diff
@@ -0,0 +1,270 @@
+	// When true, namespaces are global by default unless annotated with clustermesh.cilium.io/global=false.
+	// When false, namespaces are local by default unless annotated with clustermesh.cilium.io/global=true.
+	DefaultGlobalNamespace bool `mapstructure:"clustermesh-default-global-namespace"`
+}
+
+// Flags implements cell.Flagger to register the clustermesh-default-global-namespace flag.
+func (cfg Config) Flags(flags *pflag.FlagSet) {
+	flags.Bool("clustermesh-default-global-namespace", cfg.DefaultGlobalNamespace,
+		"Default behavior for namespaces when namespace-based filtering is active. "+
+			"When true, namespaces are global by default unless annotated with 'clustermesh.cilium.io/global=false'. "+
+			"When false, namespaces are local by default unless annotated with 'clustermesh.cilium.io/global=true'.")
+}
+
+type namespaceWatcher struct {
+	logger            *slog.Logger
+	config            Config
+	mu                lock.RWMutex
+	namespaceResource resource.Resource[*slim_corev1.Namespace]
+	nsStore           resource.Store[*slim_corev1.Namespace]
+
+	// Processors for handling namespace changes
+	processors []NamespaceProcessor
+}
+
+// NewNamespaceWatcher creates a new namespace watcher that tracks global namespaces.
+func NewNamespaceWatcher(logger *slog.Logger, config Config, namespaceResource resource.Resource[*slim_corev1.Namespace]) *namespaceWatcher {
+
+	return &namespaceWatcher{
+		logger:            logger,
+		config:            config,
+		namespaceResource: namespaceResource,
+	}
+}
+
+func (nw *namespaceWatcher) RegisterProcessor(processor NamespaceProcessor) {
+	nw.mu.Lock()
+	defer nw.mu.Unlock()
+	nw.processors = append(nw.processors, processor)
+}
+
+// isNamespaceAnnotated checks if a namespace has the global annotation by querying the resource store
+func (nw *namespaceWatcher) isNamespaceAnnotated(namespace string) bool {
+	ns, exists, err := nw.nsStore.GetByKey(resource.Key{Name: namespace})
+	if err != nil || !exists {
+		return false
+	}
+
+	_, hasAnnotation := annotation.Get(ns, annotation.GlobalNamespace)
+	return hasAnnotation
+}
+
+// isNamespaceGlobalByAnnotation checks if a namespace is marked as global by its annotation
+func (nw *namespaceWatcher) isNamespaceGlobalByAnnotation(namespace string) bool {
+	ns, exists, err := nw.nsStore.GetByKey(resource.Key{Name: namespace})
+	if err != nil || !exists {
+		return false
+	}
+
+	annotationValue, hasAnnotation := annotation.Get(ns, annotation.GlobalNamespace)
+	if !hasAnnotation {
+		return false
+	}
+
+	return strings.ToLower(annotationValue) == "true"
+}
+
+func (nw *namespaceWatcher) IsGlobalNamespace(namespace string) bool {
+	nw.mu.RLock()
+	defer nw.mu.RUnlock()
+
+	// If all namespaces are global by default (filtering inactive), then return true.
+	// Otherwise, check the annotation status.
+	return nw.config.DefaultGlobalNamespace || nw.isNamespaceGlobalByAnnotation(namespace)
```

Comment:
This logic doesn't look correct. A namespace should be global if either (a) the default is global, and the namespace is not explicitly marked as non-global via the annotation, or (b) the default is local, and the namespace is explicitly marked as global via the annotation.